timelock-server/src/main/java/com/palantir/atlasdb/timelock/AsyncTimelockServiceImpl.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock;
+
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.timelock.lock.AsyncLockService;
+import com.palantir.atlasdb.timelock.paxos.ManagedTimestampService;
+import com.palantir.lock.LockRefreshToken;
+import com.palantir.lock.v2.LockImmutableTimestampRequest;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.LockRequestV2;
+import com.palantir.lock.v2.LockTokenV2;
+import com.palantir.lock.v2.WaitForLocksRequest;
+import com.palantir.timestamp.TimestampRange;
+
+public class AsyncTimelockServiceImpl implements AsyncTimelockService {
+
+    private final AsyncLockService lockService;
+    private final ManagedTimestampService timestampService;
+
+    public AsyncTimelockServiceImpl(AsyncLockService lockService, ManagedTimestampService timestampService) {
+        this.lockService = lockService;
+        this.timestampService = timestampService;
+    }
+
+    @Override
+    public long getFreshTimestamp() {
+        return timestampService.getFreshTimestamp();
+    }
+
+    @Override
+    public TimestampRange getFreshTimestamps(int numTimestampsRequested) {
+        return timestampService.getFreshTimestamps(numTimestampsRequested);
+    }
+
+    @Override
+    public LockImmutableTimestampResponse lockImmutableTimestamp(LockImmutableTimestampRequest request) {
+        long timestamp = timestampService.getFreshTimestamp();
+
+        // this will always return synchronously
+        LockTokenV2 token = lockService.lockImmutableTimestamp(request.getRequestId(), timestamp).join();
+        long immutableTs = lockService.getImmutableTimestamp().orElse(timestamp);
```

Comment:
maybe we shouldn't call it `immutableTimestamp` on the lock service, since there's extra logic needed to determine the actual immutable timestamp if it's absent